### ISSUE TITLE
DOSBox-X Gui not accessible to screen readers, but menus do speak

### ISSUE THREAD

**Issue by blindphoneman** [joncampbell123/dosbox-x#1920]
Hi:

I was pleasantly surprised to find DOSBox-X sports a GUI menu for setting various options without editing a configuration file.

However, I tried using it with my screen-reader (NVDA) a free screen-reader from www.nvaccess.org and nothing would speak, nor did the screen reader even know it was present to interact with.

However, once I click on one of the pull-down menus, they do speak, when I use the arrow keys to go up and down, or left and right through the actual menu system (e.g. Main, CPU, Video, Sound, DOS, Capture, Drive, and Help.  I can go around through menus with arrow keys and the screen-reader speaks those with no problems.  So, apparently, standard Windows controls are being used for the menus themselves, whereas, in the dialog boxes, e.g. for the configuration, none of the buttons, combo boxes, checkboxes, etc. are identified.  This is also true with the commercial screen-readers, but, I thought I'd mention the free NVDA screen-reader, because it can be tested by anyone without any issues.  If NVDA can recognize these dialogs, I think the others would also.

Normally, a blind person can use the tab key and such to navigate the configurations, and the screen reader would tell him which one he was on, and what the current setting was.  The installer, for example, speaks very well, so I thought the rest might also speak well, but, not.

I do not mind editing the config file, so it's not a show stopper.  One thing that did have me puzzled was the "Quit Warning".  When I would get it, everything would appear to freeze from my standpoint.  The program, though was waiting for me to say that I was sure I wanted to quit.  If this had been a standard dialog box, the screen reader would have detected it and let me know when I tabbed to the appropriate button to say yes or no.  For now, I set quit warning to false.  No worse than turning off the computer prematurely back in the day, and I can definitely live with it.

Just thought I'd put it out there.

Thanks!

**Comment by joncampbell123** [comment]
Perhaps DOSBox-X could add Active Accessbility support for Windows platforms:

https://docs.microsoft.com/en-us/windows/win32/winauto/wm-getobject

Not sure how that's going to work cleanly with the SDL message architecture though. It may have to be yet another callback function pointer in the in-tree SDL1 library.

**Comment by blindphoneman** [comment]
Perhaps DOSBox-X could add Active Accessbility support for Windows platforms:
https://docs.microsoft.com/en-us/windows/win32/winauto/wm-getobject
Not sure how that's going to work cleanly with the SDL message architecture though. It may have to be yet another callback function pointer in the in-tree SDL1 library.
.

Don’t know.  I know the menus do work with no problem.

Not a show stopper, at least.
There’s also the newer thing called “UI Automation”

https://docs.microsoft.com/en-us/dotnet/framework/ui-automation/ui-automation-overview

But it may just involve more standard Windows controls.

Not a show stopper.

Reason I chose DOSBox-X is that the regular DOSBox uses some kind of faster??? Method of writing to the screen in the DOS routines (like in the command prompt), such that a DOS screen reader won’t automatically read it.  One can review the information, it’s just not read automatically.  Maybe it’s a direct write to the screen instead of through the BIOS.  DOSBox-X seems to use a more standard way of standard DOS output there and my DOS screen-reader works perfectly with it.

BTW, thanks for the additions to the Z drive.

Thanks!

**Comment by Jookia** [comment]
DOSBox-X uses a custom GUI in SDL. For it to be accessible to screen readers it would have to implement the IAccessible/AT-SPI/VoiceOver APIs on whichever platforms are available.

Which DOS screen reader are you using? Provox?

**Comment by blindphoneman** [comment]
Hi:

 

Issue 1920 relates more to Windows screen readers, as the GUI is not dos-based, is it?  If it is, that’s a whole different ballgame.

 

In the Windows environment, I use NVDA and JAWS.  NVDA is a free product, and according to it’s Web site

 

http://www.nvaccess.org

 

it is open-source.   If I could access the gui with NVDA, that would be cool.

 

Again, this is not a show-stopper.  I’m not afraid to edit a configuration file.

**Comment by Jookia** [comment]
The GUI is custom written in SDL. Only some of the GUI is in Win32, and that's on SDL1 builds.

**Comment by blindphoneman** [comment]
Hi:

In this latest build 0.837, a note in the changelog stated "  - DOSBox-X will now use native dialog box to display
    quit warnings and save state errors."

This works better already.  Before, if I got a quit warning, my Windows screen reader wouldn't tell me.  Now, Quit Warning is accessible.  This is what I'm talking about in this issue.

Could native dialogs be used for the configuration tool or other GUI functions.  Configuration tool would be very helpful.

**Comment by blindphoneman** [comment]
Forgot to say "Thanks for this change.  It is definitely appreciated.

**Comment by Wengier** [comment]
Native dialogs are mainly be used for dialogs such as message boxes, input boxes, open/save file browser. The Configuration Tool interface is quite complicated and I don't think it can be replaced with native dialogs.

**Comment by blindphoneman** [comment]
Ah, I see.  Even if it could be represented with standard dialogs, such as combo boxes, checkboxes, edit boxes, etc., it might be a terrific pain to rewrite.

I don't mind editing the .conf file since it is definitely well commented.

Could the mapper section be done as standard dialogs, or would the issues be the same?

Thanks!
